File: challenges/Services/src/number_counter_service.cpp
#include <ros/ros.h>
#include <std_msgs/Int64.h>
#include <std_srvs/SetBool.h>

int counter = 0;
ros::Publisher pub;

bool	callback_reset_counter(std_srvs::SetBool::Request &request,
				std_srvs::SetBool::Response &response)
{
	if (request.data) {
		counter = 0;
		response.success = true;
		response.message = "Counter has been reset";
	}
	else {
		response.success = false;
		response.message = "Counter has not been reset";
	}
	return true;
}

void	callback_number(const std_msgs::Int64 &msg)
{
	std_msgs::Int64 incremented_message;

	counter += msg.data;
	incremented_message.data = counter;
	pub.publish(incremented_message);
}

int	main(int argc, char **argv)
{
	ros::init(argc, argv, "Number_Counter_Service");
	ros::NodeHandle nh;
	ros::Subscriber sub = nh.subscribe("/number_publisher",
		 1000, callback_number);
	pub = nh.advertise<std_msgs::Int64>("/number_counter", 10);

	ros::ServiceServer reset_service = nh.advertiseService("/reset_counter",
		callback_reset_counter);

	ros::spin();
}

File: basics/Services/script/number_counter_service.py
#!/usr/bin/env python3

import rospy
from std_msgs.msg import Int64
from std_srvs.srv import SetBool

counter = 0
pub = None

def callback_reset_counter(req):
    if req.data:
        global counter
        counter = 0
        return True, "Counter has been reset"
    return False, "Counter has not been reset"

def number_callback(msg):
    global counter
    counter += msg.data
    newMsg = Int64()
    newMsg.data = counter
    pub.publish(newMsg)

if __name__ == '__main__':
    rospy.init_node('number_counter')

    sub = rospy.Subscriber("/number_publisher", Int64, number_callback)
    pub = rospy.Publisher("/number_counter", Int64, queue_size = 10)
    reset_service = rospy.Service("/reset_counter", SetBool, callback_reset_counter)

    rate = rospy.Rate(2)

    rospy.spin()
